Use .get() to avoid panic

jklamer opened the pull request:
# Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/3190 by avoiding panic. Close if not helpful

 # Rationale for this change
good first issue

# What changes are included in this PR?
simple api usage change. 

# Are there any user-facing changes?
solves error in scalar values less than 2 characters long I believe?

alamb reviewed (commented):
Thank you @jklamer This looks great!

Can you please add a regression test for this case? 

Perhaps you could add it along side the test added for a similar issue  https://github.com/apache/arrow-datafusion/pull/3162

kmitchener commented:
Same syntax also appears here -> https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/src/execution/context.rs#L1565

I'm not sure how that gets triggered, but it would make sense to apply the same kind of fix there also.

jklamer commented:
> Can you please add a regression test for this case?
Whoops! misread the issue and thought there was one. Adding now!

jklamer commented:
@alamb knowing where to test this was not immediately obvious to me + given there was the same issue in another place I refactored the functionality into common and tested it there. Will need to know if the test cases are accurate. Let me know what you think of this approach, I can easily walk it back. I also, cleaned up the other line mentioned in the issue but will also need some guidance where the proper place to test that is.

alamb commented on datafusion/common/src/parsing.rs:
```diff
@@ -0,0 +1,17 @@
+pub fn is_system_variables(variable_names: &[String]) -> bool {
```
FYI the reason the "RAT" test is failing on this PR is that all files need to have the Apache license at the start. 

What would you think about putting this code into the existing module https://github.com/apache/arrow-datafusion/blob/master/datafusion/physical-expr/src/var_provider.rs rather than a new module?

alamb commented on datafusion/common/src/parsing.rs:
```diff
@@ -0,0 +1,17 @@
+
+    #[test]
+    fn test_is_system_variables() {
+        assert!(!is_system_variables(&["N".into(), "\"\"".into()]));
```
Can you also add a test for a single named variable like `@F` ?

alamb commented on datafusion/core/src/execution/context.rs:
```diff
@@ -1561,8 +1561,7 @@ impl ContextProvider for SessionState {
 
-        let first_variable = &variable_names[0];
-        let provider_type = if first_variable.len() > 1 && &first_variable[0..2] == "@@" {
+        let provider_type = if is_system_variables(variable_names) {
```
This is a really nice idea to extract the common logic into a function and unit test it 🏅

alamb commented on datafusion/sql/src/planner.rs:
```diff
@@ -1737,7 +1737,7 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 let mut var_names: Vec<_> = ids.into_iter().map(|s| normalize_ident(&s)).collect();
 
-                if &var_names[0][0..1] == "@" {
+                if &var_names[0].get(0..1) == &Some("@") {
```
I am not sure it is possible to get an empty variable name here from the sqlparser

`CompoundIdentifiers` are things like `foo.bar`. I tried the following in datafusion-cli and could not provoke the issue:

```sql
❯ 
select foo.@f from foo;
SchemaError(FieldNotFound { qualifier: Some("foo"), name: "@f", valid_fields: Some(["foo.d"]) })
❯ 
select foo.@ from foo;
SchemaError(FieldNotFound { qualifier: Some("foo"), name: "@", valid_fields: Some(["foo.d"]) })
❯ 
select foo. from foo;
SchemaError(FieldNotFound { qualifier: Some("foo"), name: "from", valid_fields: Some([]) })
❯ 
select foo.d.foo from foo;
NotImplemented("Unsupported compound identifier '[\"foo\"]'")
❯ 
select foo.@.foo from foo;
NotImplemented("Unsupported compound identifier '[\"foo\"]'")
```

Thus I don't think a test is possible (and there not needed). Thank you for the change anyways

alamb reviewed (approved):
Thanks @jklamer  -- this looks great and I think is pretty much ready -- we just need a clean CI run. I suggest moving the code in datafusion/common/src/parsing.rs to datafusion/physical-expr/src/var_provider.rs which will both fix the CI and I think make for better organization

alamb reviewed (approved):
👨‍🍳 👌 

Thanks @jklamer -- really nice

andygrove commented:
@jklamer Could you run `cargo fmt` to fix the lint issue?

jklamer commented:
@andygrove absolutely. I was scared of too much reformatting before. Didn't have time last night, but will tonight

jklamer commented:
@alamb ran `cargo fmt` this time

yjshen reviewed (approved):
LGTM. Thanks @jklamer!
